cpp/src/Slice/Parser.h

```diff
@@ -356,7 +347,9 @@ namespace Slice
             KindValue
         };
 
-        Builtin(const UnitPtr& unit, Kind kind);
+        Builtin(UnitPtr unit, Kind kind);
+        void destroy();
```

Comment:
I don't understand the logic of destroy(). Which cycle are you breaking?

Reply:
Yes, using `weak_ptr` is the obvious next step in refactoring this.
And what we should of been using from the beginning.

Reply:
But just for the record, that isn't what this PR was about.
I don't want this to get beyond it's scope, and this was mostly about slimming down `SyntaxTreeBase` and untangling our inheritance hierarchy.

Although I do plan to eventually switch to using `WeakPtr` instead of this current approach of ours.